larpandoracontent/LArTwoDReco/HitWidthClusterMergingAlgorithm.cc

```diff
@@ -0,0 +1,503 @@
+    CartesianVector LSTransverseClusterFitDirection(0,0,0);
+    CartesianVector LSTransverseIntercept(0,0,0);
+    float LSTransverseChiSquared(0);
+
+    // minimise the transverse coordinate in fit (works best for longitudinal fits)
+    CartesianVector LSLongitudinalClusterFitDirection(0,0,0);
+    CartesianVector LSLongitudinalIntercept(0,0,0);
+    float LSLongitudinalChiSquared(0);
+
+    GetWeightedGradient(clusterFitParameters, true, LSTransverseClusterFitDirection, LSTransverseIntercept, LSTransverseChiSquared, fitReferencePoint);
+    GetWeightedGradient(clusterFitParameters, false, LSLongitudinalClusterFitDirection, LSLongitudinalIntercept, LSLongitudinalChiSquared, fitReferencePoint);
+
+    // return fit with the lowest chi-squared
+    if(LSTransverseChiSquared < LSLongitudinalChiSquared)
+        return LSTransverseIntercept;
+
+    return LSLongitudinalIntercept;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+
+void HitWidthClusterMergingAlgorithm::GetWeightedGradient(const LArHitWidthHelper::ClusterParameters &clusterFitParameters, bool isTransverse, CartesianVector &direction, CartesianVector &intercept, float &chiSquared, const CartesianVector &fitReferencePoint) const
+{
+
+    // cannot make a longitudinal fit to a single hit cluster
+    if(!isTransverse && clusterFitParameters.GetNumCaloHits() == 1) 
+    {
+        std::cout << "WARNING - CANNOT MAKE LONGITUDINAL FIT TO SINGLE HIT CLUSTER, RETURNED" << std::endl;
+        throw;
+    }
+
+    float weightSum(0);
+    float weightedXSum(0);
+    float weightedZSum(0);
+
+    bool isXConstant(true);
+    bool isZConstant(true);
+
+    LArHitWidthHelper::ConstituentHitVector constituentHitVector(clusterFitParameters.GetConstituentHitVector());
+
+    // sort hits with respect to their distance to the fitReferencePoint (closest -> furthest)
+    std::sort(constituentHitVector.begin(), constituentHitVector.end(), LArHitWidthHelper::ConstituentHit::SortByDistanceToPoint(fitReferencePoint));
+
+    for(const LArHitWidthHelper::ConstituentHit &constituentHit : constituentHitVector) 
+    {
+        const CartesianVector hitPosition = constituentHit.GetPositionVector();
+        const float hitWeight = constituentHit.GetHitWidth();
+
+        if(std::fabs(clusterFitParameters.GetConstituentHitVector().begin()->GetPositionVector().GetX() - hitPosition.GetX()) > std::numeric_limits<float>::epsilon())
+            isXConstant = false;
+
+        if(std::fabs(clusterFitParameters.GetConstituentHitVector().begin()->GetPositionVector().GetZ() - hitPosition.GetZ()) > std::numeric_limits<float>::epsilon())
+            isZConstant = false;
+
+        weightedXSum += hitPosition.GetX() * hitWeight;
+        weightedZSum += hitPosition.GetZ() * hitWeight;
+        weightSum += hitWeight;
+        
+        if(weightSum > m_fittingWeight)
```

Comment:
in this function i am performing a least squared fit. In the first for loop i am finding the weighted X and Z mean, in the second loop i find the gradient and z-intercept and in the final for loop i calculate the chi-squared for the fit. But i also want to stop this process in some cases and i use the first loop to find these cases - thats what lines 378-382 are about. I am trying to stop the fit when X/Z is constant as these are just horizontal/vertical fits and will result in dividing by zero later on in the function. 

To reduce computational times I also only include a certain amount of hit weight in the fit and thats what the weightSum > m_fittingWeight statement is about. Clearly, i only want to include the hits closest to the merge point in the fit and so that what the sorting in line 371 is about. The reference point is the lowerXExtrema for a considered cluster and the higherXExtrema for a current cluster. 

Also, I should add, that I have implemented the sliding linear fitting but i find that they never really work and i think this is due to the smallness in the layers occupied by the hits of the cluster or the smallness in the number of hits in the cluster. So I rely on my cluster fitting function and I think it does a pretty good job.

But I will add comments to the function to explain what is happening

Reply:
Overview would go nicely in the doxygen @brief description for this function, with additional one-line clarifications where the variable names/function calls etc. can't make it sufficiently clear what's going on. Trying to keep the number of lines down as much as possible (so e.g. creating multiple variables, such as your three floats two Booleans, on one line is fine - but never any operations on variables: one operation per line), so long as not making things too terse to follow.